### PR TITLE
Infer Share

### PR DESCRIPTION
**Pull request by calebmkim**
This is a continuation of domination_map stuff, but I b/c I messed up using git the easiest thing was to create a new branch. 

I now have added an infer-share pass that attaches the "state_share" attribute to components that are state_shareable. 
Like before, it may be easier to talk in person about what's going on in the code. 

It first looks through for any a) user-defined, non shareable components b) any mem cells and c) any ref cells and automatically labels the component "not shareable" if it uses any of those things. Then it builds a domination map, and for each (node, dominators of node) pairing in the map it looks through node and finds all the reads to state_shareable components in the node (i.e. a control statement)  and makes sure it is dominated by a write to that component. One thing I had to tweak a little bit is that you can't consider a node as one of its own dominators, since very often a read/write in the same group will *not* be state shareable. 

I also factored out the "share_set" code bc I found `infer-share` wanted to use it too.

### CODE REVIEW

File: calyx/src/analysis/share_set.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+impl ShareSet {
+    pub fn new(shareable: HashSet<ir::Id>, is_state_share: bool) -> Self {
+        ShareSet { shareable, is_state_share  }
+    }
+
+    ///Constructs a shareset from the context. Looks for "state_share" types if 
+    ///is_state_share is true, and "share" types otherwise. 
+    pub fn from_context(ctx: &ir::Context, is_state_share: bool) -> Self {
```
Comment (rachitnigam) [inline calyx/src/analysis/share_set.rs:23]: One trick to use here is you expect `is_state_share` to always be a constant (either `true` or `false`) is to make the function use [`const` generics](https://blog.rust-lang.org/2021/02/26/const-generics-mvp-beta.html):
```
pub fn from_context<const IS_STATE_SHARE: bool>(ctx: &ir::Context) { ... }
```

You can use `IS_STATE_SHARE` normal in the function body and at call sites, you can do:
```
from_context::<true>(ctx)
from_context::<false>(ctx)
```

The benefit of this is the compiler will generate a specialized version for each value of `IS_STATE_SHARE` for you

File: calyx/src/passes/infer_share.rs [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+use crate::ir::CloneName;
+use std::collections::HashSet;
+use std::{rc::Rc};
+
+const BEGIN_ID: &str = "BEGIN_ID";
+const END_ID: &str = "END_ID";
+
+//given an RRC of a cell, determines if cell is a std_mem_cell. There may be an 
+//easier way to do this.
+fn is_std_mem(cell: &ir::Cell) -> bool{
```
Comment (rachitnigam) [inline calyx/src/passes/infer_share.rs:16]: Why are memories specially defined to be unshareable? They also don't have the `state_share` attribute right?

File: tests/passes/infer-share-dom-map/dmap.expect [hunk 3]
```diff
@@ -0,0 +1,157 @@
+Domination Map {
```
Comment (rachitnigam) [inline tests/passes/infer-share-dom-map/dmap.expect:1]: Did you mean to print out the Domination Map here?

**Comment by calebmkim** [comment]
Sure, but just so I'm clear, do you want me to clarify
 a) how I am attaching the BEGIN_ID and NODE_ID attribute to the control statements 
or 
b) after I have attached BEGIN_ID and NODE_ID to control statements, how I am using them in my program?

File: tests/passes/infer-share-dom-map/dmap.expect (hunk 3 above)
Comment (calebmkim) [reply tests/passes/infer-share-dom-map/dmap.expect:1]: Right now, I have 2 groups of tests for the infer-share pass: the first prints out the domination map to make sure it's correct, the second tests whether the infer-share pass correctly attaches the "state_share" attribute or not. My thinking behind this was that the first is mainly testing whether domination map analysis is correct, and the second is testing whether my infer-share pass is correct (although this pass obviously relies on my domination map being correct)

Comment (rachitnigam) [reply tests/passes/infer-share-dom-map/dmap.expect:1]: Okay, if you're going to print out the dominationMap then let's make it a little more human-readable than it is right now. For example, to an external user, these numbers don't obviously correspond to anything. Furthermore, we use the [`get_opts`](https://github.com/cucapra/calyx/blob/master/calyx/src/passes/top_down_compile_control.rs#L612) method to create a new option for the `infer-share` pass to dump out this information when the user provides a CLI option.

**Comment by rachitnigam** [comment]
@calebmkim it seems like this PR has changes to both `cell_share` and the new infer cell sharing stuff. Can you break it up into two PRs, one which makes the changes to cell_share that don't effect infer cell sharing and one with the new infer sharing stuff.

You can take advantage of `git cherry-pick` to copy specific commits on this branch to a new one. In general, PRs should have the minimal number of changes to make them easy to review and merge. Additionally, because we have a "squash and merge" policy (which squashes all commits in a PR into one commit), it is doubly useful to have smaller PRs that can be merge independently.

For example, we expect the changes to `cell_share` to not effect anything behaviorally while we do expect things to change when we merge the `infer_cell_share` pass.

File: tests/passes/infer-share-dom-map/dmap.expect (hunk 3 above)
Comment (calebmkim) [reply tests/passes/infer-share-dom-map/dmap.expect:1]: I can' think of something else other than numbers to print to the d_map. 
Printing the numbers makes the most sense to me; but maybe when I print it I can explain that the numbers correspond to the NODE_ID, BEGIN_ID, and END_ID attributes of the control statements (which are visible at the end of the file)? 
Furthermore, I should definitely print the name of the component of the domination map that I am printing. 
In addition, I can print it a bit more nicely so the user can see very easily which nodes are dominated by which? 
If you have a different suggestion for how I should approach this then lmk.

**Comment by rachitnigam** [comment]
Does #1070 subsume this? If so, let's close this

**Comment by calebmkim** [comment]
Ok, closing. This PR has been split into #1069 and #1070.

**Pull request closed without merging**